src/Development/IDE/Core/FileExists.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE OverloadedLists      #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Development.IDE.Core.FileExists
+  ( fileExistsRules
+  , modifyFileExists
+  , getFileExists
+  )
+where
+
+import           Control.Concurrent.Extra
+import           Control.Exception
+import           Control.Monad
+import qualified Data.Aeson                    as A
+import           Data.Binary
+import qualified Data.ByteString.Lazy          as BS
+import           Data.Map.Strict                ( Map )
+import qualified Data.Map.Strict               as Map
+import qualified Data.Text                     as T
+import           Development.IDE.Core.Shake
+import           Development.IDE.Types.Location
+import           Development.Shake
+import           Development.Shake.Classes
+import           GHC.Generics
+import           Language.Haskell.LSP.Messages
+import           Language.Haskell.LSP.Types
+import           Language.Haskell.LSP.Types.Capabilities
+
+-- | A map for tracking the file existence
+type FileExistsMap = (Map NormalizedFilePath Bool)
+
+-- | A wrapper around a mutable 'FileExistsMap'
+newtype FileExistsMapVar = FileExistsMapVar (Var FileExistsMap)
+
+instance IsIdeGlobal FileExistsMapVar
+
+-- | Grab the current global value of 'FileExistsMap' without acquiring a dependency
+getFileExistsMapUntracked :: Action FileExistsMap
+getFileExistsMapUntracked = do
+  FileExistsMapVar v <- getIdeGlobalAction
+  liftIO $ readVar v
+
+-- | Modify the global store of file exists
+modifyFileExistsAction :: (FileExistsMap -> IO FileExistsMap) -> Action ()
+modifyFileExistsAction f = do
+  FileExistsMapVar var <- getIdeGlobalAction
+  liftIO $ modifyVar_ var f
+
+-- | Modify the global store of file exists
+modifyFileExists :: IdeState -> [(NormalizedFilePath, Bool)] -> IO ()
+modifyFileExists state changes = do
+  FileExistsMapVar var <- getIdeGlobalState state
+  changesMap           <- evaluate $ Map.fromList changes
+  modifyVar_ var $ evaluate . Map.union changesMap
```

Comment:
This looks like it might need to be masked. Otherwise, you can end up modifying `changesMap` without deleting the values if you get interrupted in between.

Reply:
Thanks, I didn't think of that. In order to delete the values from the state though, I need to call `runShake`. Is it safe to call `runShake` masked?

Reply:
`runShake` has a nontrivial overhead so given that we only want to modify the var here, maybe it would make sense to refactor this such that we don’t need `runShake` here at all, e.g., by allocating the Var in `initialise` and then passing it on.

Reply:
Does the overhead matter if we only subscribe to create and delete events?

Is that refactoring small enough to include in this PR?

Reply:
I think it makes sense to include it in this PR mainly because I’m not entirely comfortable with masking `shakeRun`. As far as I can tell, the only reason you need to do that is to get access to `ShakeExtras` which is already in `IdeState` so if you simply change `deleteValue` to take `IdeState` as the input, you should be able to avoid this.